zetaclient/chains/bitcoin/observer/inbound.go

```diff
@@ -472,3 +472,173 @@ func GetBtcEvent(
+	netParams *chaincfg.Params,
+	depositorFee float64,
+) (*BTCInboundEvent, error) {
+	if len(tx.Vout) < 1 {
+		logger.Debug().Msgf("no output %s", tx.Txid)
+		return nil, nil
+	}
+	if len(tx.Vin) == 0 { // should never happen
+		logger.Debug().Msgf("no input found for inbound: %s", tx.Txid)
+		return nil, nil
+	}
+
+	if !isValidRecipient(tx.Vout[0].ScriptPubKey.Hex, tssAddress, netParams, logger) {
+		logger.Debug().Msgf("irrelevant recipient %s for inbound: %s", tx.Vout[0].ScriptPubKey.Hex, tx.Txid)
+		return nil, nil
+	}
+
+	isAmountValid, amount := isValidAmount(tx.Vout[0].Value, depositorFee)
+	if !isAmountValid {
+		logger.Info().
+			Msgf("GetBtcEventWithWitness: btc deposit amount %v in txid %s is less than depositor fee %v", tx.Vout[0].Value, tx.Txid, depositorFee)
+		return nil, nil
+	}
+
+	var memo []byte
+	if candidate := tryExtractOpRet(tx, logger); candidate != nil {
+		memo = candidate
+		logger.Debug().Msgf("GetBtcEventWithWitness: found OP_RETURN memo %s in tx %s", hex.EncodeToString(memo), tx.Txid)
+	} else if candidate = tryExtractInscription(tx, logger); candidate != nil {
+		memo = candidate
+		logger.Debug().Msgf("GetBtcEventWithWitness: found inscription memo %s in tx %s", hex.EncodeToString(memo), tx.Txid)
+	} else {
+		return nil, nil
+	}
+
+	// event found, get sender address
+	fromAddress, err := GetSenderAddressByVin(client, tx.Vin[0], netParams)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error getting sender address for inbound: %s", tx.Txid)
+	}
+
+	return &BTCInboundEvent{
+		FromAddress: fromAddress,
+		ToAddress:   tssAddress,
+		Value:       amount,
+		MemoBytes:   memo,
+		BlockNumber: blockNumber,
+		TxHash:      tx.Txid,
+	}, nil
+}
+
+func tryExtractOpRet(tx btcjson.TxRawResult, logger zerolog.Logger) []byte {
+	if len(tx.Vout) < 2 {
+		logger.Debug().Msgf("txn %s has fewer than 2 outputs, not target OP_RETURN txn", tx.Txid)
+		return nil
+	}
+
+	vout1 := tx.Vout[1]
+	memo, found, err := bitcoin.DecodeOpReturnMemo(vout1.ScriptPubKey.Hex, tx.Txid)
+	if err != nil {
+		logger.Error().Err(err).Msgf("tryExtractOpRet: error decoding OP_RETURN memo: %s", vout1.ScriptPubKey.Hex)
+		return nil
+	}
+
+	if found {
+		return memo
+	}
+	return nil
+}
+
+// ParseScriptFromWitness attempts to parse the script from the witness data. Ideally it should be handled by
+// bitcoin library, however, it's not found in existing library version. Replace this with actual library implementation
+// if libraries are updated.
+func ParseScriptFromWitness(witness []string, logger zerolog.Logger) []byte {
+	length := len(witness)
+
+	if length == 0 {
+		return nil
+	}
+
+	lastElement, err := hex.DecodeString(witness[length-1])
+	if err != nil {
+		logger.Debug().Msgf("invalid witness element")
+		return nil
+	}
+
+	// From BIP341:
+	// If there are at least two witness elements, and the first byte of
+	// the last element is 0x50, this last element is called annex a
+	// and is removed from the witness stack.
```

Comment:
nit: really curious, what's the use of annex?

Reply:
From the spec, it's really reserved for future changes. I dont think it's use for now.